### PR TITLE
Refactor vision tasks(image classification, object detection)

### PR DESCRIPTION
**Pull request by zhreshold**
*Description of changes:*
- Refactor vision tasks(image classification, object detection to align with new public APIs)
- split jenkins unittests for each module, let them run on separate nodes to speed up unittests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-733-30 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-733/30/index.html

**Comment by Innixma** [comment]
Cool stuff regarding splitting the jenkins unittests to be parallel. Do you know if this is possible for the doc build?

**Comment by zhreshold** [comment]
> Cool stuff regarding splitting the jenkins unittests to be parallel. Do you know if this is possible for the doc build?

For doc it's not possible yet, as the built results must be collected by sphinx and generate html after all tutorials are complete. I heard there are some advanced artifacts sharing techniques across different nodes, but aren't quite sure about that.

File: Jenkinsfile [hunk 1]
```diff
@@ -1,74 +1,275 @@
...
-        set -ex
-        conda env update -n autogluon_py3 -f docs/build.yml
-        conda activate autogluon_py3
-        conda list
-        export CUDA_VISIBLE_DEVICES=${VISIBLE_GPU}
-        env
-        export LD_LIBRARY_PATH=/usr/local/cuda-10.1/lib64
-        export MPLBACKEND=Agg
-        export MXNET_CUDNN_AUTOTUNE_DEFAULT=0
+  parallel 'core': {
```
Comment (Innixma) [inline Jenkinsfile:4]: Since this is a really nice change that is separate from the vision task refactor logic, could this be moved into a separate PR so we can merge it in quickly?

Comment (zhreshold) [reply Jenkinsfile:4]: yes, see https://github.com/awslabs/autogluon/pull/773

**Comment by szha** [comment]
Job PR-733-40 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-733/40/index.html

**Comment by szha** [comment]
Job PR-733-45 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-733/45/index.html

**Comment by szha** [comment]
Job PR-733-47 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-733/47/index.html

**Comment by szha** [comment]
Job PR-733-48 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-733/48/index.html

**Comment by zhreshold** [comment]
Ready for review, with docs(http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-733/48/index.html)
@Innixma @jwmueller @sxjscience @gradientsky Appreciated!

**Review by gradientsky — APPROVED** [review]

**Review by sxjscience — APPROVED** [review]

File: docs/tutorials/image_classification/beginner.md [hunk 2]
```diff
@@ -63,36 +42,63 @@ The best model is finally retrained on our entire dataset (i.e., merging trainin
...
+print(result)
+```
 
+You can also feed in multiple images all together, let's use images in test dataset as an example:
 ```{.python .input}
-# skip this if training FashionMNIST on CPU.
-if ag.get_gpu_count() > 0:
-    image = 'data/test/BabyShirt/BabyShirt_323.jpg'
-    ind, prob, _ = classifier.predict(image, plot=True)
+bulk_result = classifier.predict(test_dataset)
```
Comment (Innixma) [inline docs/tutorials/image_classification/beginner.md:62]: Looking at the tutorial, I notice that the batch predict output differs from tabular, it looks like you have 1 row per class per image. Is this what users ideally want? If they wanted this format, wouldn't it make more sense to be in predict_proba, with N columns for N classes, each with the proba (as in Tabular)? This would keep the rows equal to the number of images, and allow for the index to remain aligned.

Would there be any drawbacks to having predict simply return the name of the majority class (as done in Tabular)?

**Review by Innixma — COMMENTED** [review]
Will plan to do a review Tuesday.

File: docs/tutorials/image_classification/beginner.md (hunk 2 above)
Comment (zhreshold) [reply docs/tutorials/image_classification/beginner.md:62]: Okay, I will try keep it aligned with tabular predictor, for single input, return the category id; for multiple inputs, 1 row per image. The current `predict` is actually `predict_proba` in the context of tabular predictor.

Comment (zhreshold) [reply docs/tutorials/image_classification/beginner.md:62]: *correction, for `predict_proba` it still needs to group results by image

File: vision/tests/unittests/test_object_detection.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+def test_task():
+    dataset = Task.Dataset.from_voc('https://autogluon.s3.amazonaws.com/datasets/tiny_motorbike.zip')
+    train_data, _, test_data = dataset.random_split()
+
+    detector = Task()
+    detector.fit(train_data, epochs=1, num_trials=1, hyperparameters={'batch_size': 4})
+    test_result = detector.predict(test_data)
+    print('test result', test_result)
+    detector.save('detector.ag')
+    detector2 = Task.load('detector.ag')
```
Comment (Innixma) [inline vision/tests/unittests/test_object_detection.py:12]: confirm that predictions pre and post loading are identical?

File: vision/tests/unittests/test_image_classification.py [hunk 4]
```diff
@@ -1,24 +1,14 @@
...
+def test_task():
+    dataset, _, test_dataset = Task.Dataset.from_folders('https://autogluon.s3.amazonaws.com/datasets/shopee-iet.zip')
+    model_list = Task.list_models()
+    classifier = Task()
+    classifier.fit(dataset, epochs=1, num_trials=2)
+    test_result = classifier.predict(test_dataset)
+    print('test result', test_result)
     classifier.save('classifier.ag')
-    classifier2 = task.Classifier.load('classifier.ag')
+    classifier2 = Task.load('classifier.ag')
```
Comment (Innixma) [inline vision/tests/unittests/test_image_classification.py:11]: confirm that predictions pre and post loading are identical?

File: vision/setup.py [hunk 5]
```diff
@@ -48,9 +48,10 @@ def create_version_file():
     'Pillow<=6.2.1',
     'numpy>=1.16.0',
     'matplotlib',
-    'gluoncv>=0.5.0,<1.0',
+    'gluoncv==0.9.0b20201125',
```
Comment (Innixma) [inline vision/setup.py:51]: What is the planned version of gluoncv we will officially be using for v0.1 release? I'd prefer to not depend on a pre-release build.

File: vision/setup.py [hunk 6]
```diff
@@ -48,9 +48,10 @@ def create_version_file():
...
     'graphviz<0.9.0,>=0.8.1',
     'pandas>=1.0.0,<2.0',
+    'd8>=0.0.2',
```
Comment (Innixma) [inline vision/setup.py:54]: Please cap version such as <0.1 or <1 , since its possible API could break given it is so new.

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 7]
```diff
@@ -1,341 +1,256 @@
...
+import logging
+
+import pandas as pd
+from gluoncv.auto.tasks import ImageClassification as _ImageClassification
+from gluoncv.model_zoo import get_model_list
 
 __all__ = ['ImageClassification']
 
-from autogluon.core import sample_config
+class ImageClassification(object):
```
Comment (Innixma) [inline vision/src/autogluon/vision/image_classification/image_classification.py:12]: `ImageClassifier`? `ImagePredictor`? `ImageClassification` doesn't seem like the right name for a class. I personally think `ImagePredictor` is best as it leaves the door open to expand to regression, and it will align with `TabularPredictor`.

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 8]
```diff
@@ -1,341 +1,256 @@
...
+    def __init__(self, log_dir=None):
+        self._log_dir = log_dir
+        self._classifier = None
+        self._fit_summary = {}
 
-class ImageClassification(BaseTask):
-    """AutoGluon Task for classifying images based on their content
-    """
-    Classifier = Classifier
+    def fit(self,
```
Comment (Innixma) [inline vision/src/autogluon/vision/image_classification/image_classification.py:33]: No eval_metric?

Comment (Innixma) [inline vision/src/autogluon/vision/image_classification/image_classification.py:33]: No verbosity control?

File: docs/tutorials/object_detection/beginner.md [hunk 9]
```diff
@@ -5,81 +5,68 @@ Object detection is the process of identifying and localizing objects in an imag
...
-time_limits = 5*60*60  # 5 hours
-epochs = 30
-detector = task.fit(dataset_train,
-                    num_trials=2,
-                    epochs=epochs,
-                    lr=ag.Categorical(5e-4, 1e-4),
-                    ngpus_per_trial=1,
-                    time_limits=time_limits)
+time_limit = 60*60  # 1 hour
+task = Task()
```
Comment (Innixma) [inline docs/tutorials/object_detection/beginner.md:34]: `detector = ObjectDetector()` ?

File: docs/tutorials/object_detection/beginner.md [hunk 10]
```diff
@@ -5,81 +5,68 @@ Object detection is the process of identifying and localizing objects in an imag
...
+detector = task.fit(dataset_train, time_limit=time_limit, num_trials=2, epochs=5)
```
Comment (Innixma) [inline docs/tutorials/object_detection/beginner.md:35]: `detector.fit(dataset_train, time_limit=time_limit, num_trials=2, epochs=5)` with no return seems cleaner.

File: docs/tutorials/image_classification/kaggle.md [hunk 11]
```diff
@@ -15,7 +15,7 @@ After completing this step, you will have the following directory structure on y
     ├── ...
 ```
 
-Here `data` is a folder containing the raw images categorized into classes. For example, subfolder `class1` contains all images that belong to the first class, `class2` contains all images belonging to the second class, etc. 
+Here `data` is a folder containing the raw images categorized into classes. For example, subfolder `class1` contains all images that belong to the first class, `class2` contains all images belonging to the second class, etc.
```
Comment (Innixma) [inline docs/tutorials/image_classification/kaggle.md:18]: Where is the kaggle tutorial in the doc build? I can't find it.

File: docs/api/autogluon.task.rst [hunk 12]
```diff
@@ -9,22 +9,20 @@ AutoGluon Tasks
...
    >>> import autogluon.core as ag
-   >>> from autogluon.vision import ImageClassification as task
+   >>> from autogluon.vision import ImageClassification as Task
 
    Load a toy image dataset:
 
-   >>> filename = ag.download('http://autogluon-hackathon.s3.amazonaws.com/data.zip')
-   >>> ag.unzip(filename)
-   >>> dataset = task.Dataset(train_path='data/train')
+   >>> train, val, test = Task.Dataset.from_folders('http://autogluon-hackathon.s3.amazonaws.com/data.zip')
```
Comment (Innixma) [inline docs/api/autogluon.task.rst:16]: nit: probably want to move datasets in the hackathon S3 bucket to the main `autogluon` S3 bucket to avoid inconsistencies.

**Review by Innixma — CHANGES_REQUESTED** [review]

File: vision/src/autogluon/vision/object_detection/object_detection.py [hunk 13]
```diff
@@ -1,303 +1,175 @@
...
-from autogluon.core.decorator import sample_config
-from autogluon.core.space import Categorical
-from autogluon.core.utils import get_cpu_count, get_gpu_count
-from autogluon.mxnet.utils import update_params
+from gluoncv.auto.tasks import ObjectDetection as _ObjectDetection
 
 __all__ = ['ObjectDetection']
 
-logger = logging.getLogger(__name__)
+class ObjectDetection(object):
```
Comment (Innixma) [inline vision/src/autogluon/vision/object_detection/object_detection.py:9]: Ditto to the majority of my comments on ImageClassification.

**Comment by szha** [comment]
Job PR-733-49 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-733/49/index.html

File: docs/tutorials/image_classification/beginner.md [hunk 14]
```diff
@@ -18,42 +18,21 @@ For demonstration purposes, we use a subset of the [Shopee-IET dataset](https://
...
 Now, we fit a classifier using AutoGluon as follows:
 
 ```{.python .input}
-classifier = task.fit(dataset,
-                      epochs=5,
-                      ngpus_per_trial=1,
-                      verbose=False)
+task = Task()
+# since the original dataset does not provide validation split, the `fit` function splits it randomly with 90/10 ratio
+classifier = task.fit(train_dataset, epochs=2)  # you can trust the default config, we reduce the # epoch to save some build time
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply docs/tutorials/image_classification/beginner.md:35]: there's no specific reason why it can't, the I can use `classifier = xxx()` and `classifier.fit()` without returned value. Updated for object detection as well

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 15]
```diff
@@ -1,341 +1,256 @@
...
 __all__ = ['ImageClassification']
 
-from autogluon.core import sample_config
+class ImageClassification(object):
+    """AutoGluon Predictor for classifying images based on their whole contents
 
-logger = logging.getLogger(__name__)
+    Parameters
+    ----------
+    config : dict
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:17]: Fixed

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 16]
```diff
@@ -1,341 +1,256 @@
...
+        The configurations, can be nested dict.
+    logger : logging.Logger
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py]: Fixed

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 17]
```diff
@@ -1,341 +1,256 @@
...
+        The desired logger object, use `None` for module specific logger with default setting.
+    net : mx.gluon.Block
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py]: Fixed

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 18]
```diff
@@ -1,341 +1,256 @@
...
+        The custom network. If defined, the model name in config will be ignored so your
+        custom network will be used for training rather than pulling it from model zoo.
+    """
+    # Dataset is a subclass of `pd.DataFrame`, with `image` and `label` columns.
+    Dataset = _ImageClassification.Dataset
 
+    def __init__(self, log_dir=None):
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:24]: Fixed

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 19]
```diff
@@ -1,341 +1,256 @@
...
+            If `None`, the validation dataset will be randomly split from `train_data`.
+        train_size : float
+            The random split ratio for `train_data` if `val_data==None`.
+            The new `val_data` size will be `1-train_size`.
+        random_state : numpy.random.state
+            The random_state for shuffling, only used if `val_data==None`.
+            Note that the `random_state` only affect the splitting process, not model training.
+        time_limit : int
+            Time limit in seconds, default is 12 hours. If `time_limit` is hit during `fit`, the
+            HPO process will interupt and return the current best configuration.
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:63]: Fixed

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 20]
```diff
@@ -1,341 +1,256 @@
...
+            Training data, can be a dataframe like image dataset.
+            If a string is provided, will search for k8 datasets.
+            If `None`, the validation dataset will be randomly split from `train_data`.
+        train_size : float
+            The random split ratio for `train_data` if `val_data==None`.
+            The new `val_data` size will be `1-train_size`.
+        random_state : numpy.random.state
+            The random_state for shuffling, only used if `val_data==None`.
+            Note that the `random_state` only affect the splitting process, not model training.
+        time_limit : int
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:61]: Added

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 21]
```diff
@@ -1,341 +1,256 @@
...
+            Training data, can be a dataframe like image dataset.
+            If a string is provided, will search for k8 datasets.
+        val_data : pd.DataFrame or str
+            Training data, can be a dataframe like image dataset.
+            If a string is provided, will search for k8 datasets.
+            If `None`, the validation dataset will be randomly split from `train_data`.
+        train_size : float
+            The random split ratio for `train_data` if `val_data==None`.
+            The new `val_data` size will be `1-train_size`.
+        random_state : numpy.random.state
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:58]: default = None, added

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 22]
```diff
@@ -1,341 +1,256 @@
...
+            The random_state for shuffling, only used if `val_data==None`.
+            Note that the `random_state` only affect the splitting process, not model training.
+        time_limit : int
+            Time limit in seconds, default is 12 hours. If `time_limit` is hit during `fit`, the
+            HPO process will interupt and return the current best configuration.
+        epochs : int
+            The `epochs` for model training, if `None` is provided, then default `epochs` for model
+            will be used.
+        num_trials : int, default is 1
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:67]: fixed

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 23]
```diff
@@ -1,341 +1,256 @@
...
+        Parameters
+        ----------
+        train_data : pd.DataFrame or str
+            Training data, can be a dataframe like image dataset.
+            If a string is provided, will search for k8 datasets.
+        val_data : pd.DataFrame or str
+            Training data, can be a dataframe like image dataset.
+            If a string is provided, will search for k8 datasets.
+            If `None`, the validation dataset will be randomly split from `train_data`.
+        train_size : float
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:55]: I am changing it to `holdout_frac` for consistency, could update to `val_frac` all together with tabular

File: vision/src/autogluon/vision/image_classification/image_classification.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:58]: I am following the numpy random state definition: https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.random.RandomState.html#numpy.random.RandomState, it supports non-integer value.

I have no preference for `random_seed` or `random_state`

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 24]
```diff
@@ -1,341 +1,256 @@
...
+            Note that the `random_state` only affect the splitting process, not model training.
+        time_limit : int
+            Time limit in seconds, default is 12 hours. If `time_limit` is hit during `fit`, the
+            HPO process will interupt and return the current best configuration.
+        epochs : int
+            The `epochs` for model training, if `None` is provided, then default `epochs` for model
+            will be used.
+        num_trials : int, default is 1
+            The number of HPO trials. If `None`, will run only one trial.
+        hyperparameters : dict
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:69]: Fixed

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 25]
```diff
@@ -1,341 +1,256 @@
...
+            The random_state for shuffling, only used if `val_data==None`.
+            Note that the `random_state` only affect the splitting process, not model training.
+        time_limit : int
+            Time limit in seconds, default is 12 hours. If `time_limit` is hit during `fit`, the
+            HPO process will interupt and return the current best configuration.
+        epochs : int
+            The `epochs` for model training, if `None` is provided, then default `epochs` for model
+            will be used.
+        num_trials : int, default is 1
+            The number of HPO trials. If `None`, will run only one trial.
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:68]: ya, this one sounds good to me. I am changing to infinity runs for `None`

[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:68]: It will sample from search space if it indeed a search space, otherwise the default is used

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 26]
```diff
@@ -1,341 +1,256 @@
...
+        hyperparameters : dict
+            Extra hyperparameters for specific models.
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:70]: I will include a few important hps here, the full list is > 100, a valid list can be viewed in a generated config.yaml file

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 27]
```diff
@@ -1,341 +1,256 @@
...
+        search_strategy : str
+            Searcher strategy for HPO, 'random' by default.
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:84]: added

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 28]
```diff
@@ -1,341 +1,256 @@
...
+        scheduler_options : dict
+            Extra options for HPO scheduler, please refer to `autogluon.Searcher` for details.
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:86]: Fixed

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 29]
```diff
@@ -1,341 +1,256 @@
...
+                         'rec_val : ~/.mxnet/datasets/imagenet/rec/val.rec\n' +
+                         'rec_val_idx : ~/.mxnet/datasets/imagenet/rec/val.idx\n')
+            train_data = pd.DataFrame({'image': [], 'label': []})
+            val_data = pd.DataFrame({'image': [], 'label': []})
+            use_rec = True
+        if isinstance(train_data, str):
+            from d8.image_classification import Dataset as D8D
+            names = D8D.list()
+            if train_data.lower() in names:
+                train_data = D8D.get(train_data)
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:115]: only applies to datasets as str: e.g., cifar10,

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 30]
```diff
@@ -1,341 +1,256 @@
...
+            if val_data is None:
+                train_data, val_data = train_data.split(train_size)
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:117]: Fixed

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 31]
```diff
@@ -1,341 +1,256 @@
...
+        if isinstance(val_data, str):
+            from d8.image_classification import Dataset as D8D
+            names = D8D.list()
+            if val_data.lower() in names:
+                val_data = D8D.get(val_data)
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:122]: Fixed

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 32]
```diff
@@ -1,341 +1,256 @@
...
+        if self._classifier is not None:
+            self._fit_summary = self._classifier.fit(train_data, val_data, train_size, random_state, resume=False)
+            return
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:126]: return self will be compatible with previous `xxx = task.fit()` usage, though not necessary any more

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 33]
```diff
@@ -1,341 +1,256 @@
...
+                train_data = D8D.get(train_data)
+            if val_data is None:
+                train_data, val_data = train_data.split(train_size)
+        if isinstance(val_data, str):
+            from d8.image_classification import Dataset as D8D
+            names = D8D.list()
+            if val_data.lower() in names:
+                val_data = D8D.get(val_data)
+        if self._classifier is not None:
+            self._fit_summary = self._classifier.fit(train_data, val_data, train_size, random_state, resume=False)
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:125]: yep, this is exactly doing the `fit_extra` work with very limited argument options, and will overwrite the previous model

File: vision/src/autogluon/vision/image_classification/image_classification.py (hunk 8 above)
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:29]: added verbosity similar to tabular

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 34]
```diff
@@ -1,341 +1,256 @@
...
+            optimizer = hyperparameters.pop('optimizer', None)
+            if optimizer is not None:
+                config.update({'custom_optimizer': optimizer})
+            config.update(hyperparameters)
+        if scheduler_options is not None:
+            config.update(scheduler_options)
+        if use_rec == True:
+            config['use_rec'] = True
+        task = _ImageClassification(config=config)
+        self._classifier = task.fit(train_data, val_data, train_size, random_state)
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:156]: we didn't split pandas compatible datasets, this will be verified again so it won't do double splitting

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 35]
```diff
@@ -1,341 +1,256 @@
...
+            return
+
+        # new HPO task
+        config={'log_dir': self._log_dir,
+                'num_trials': 1 if num_trials is None else num_trials,
+                'time_limits': time_limit,
+                'search_strategy': search_strategy,
+                }
+        if nthreads_per_trial is not None:
+            config.update({'nthreads_per_trial': nthreads_per_trial})
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:135]: checking for `None` will avoid adding new keys to config, reducing the args passed to searcher as well

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 36]
```diff
@@ -1,341 +1,256 @@
...
+        if epochs is not None:
+            config.update({'epochs': epochs})
+        if isinstance(hyperparameters, dict):
+            net = hyperparameters.pop('net', None)
+            if net is not None:
+                config.update({'custom_net': net})
+            optimizer = hyperparameters.pop('optimizer', None)
+            if optimizer is not None:
+                config.update({'custom_optimizer': optimizer})
+            config.update(hyperparameters)
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:149]: sure, added a checker

File: vision/src/autogluon/vision/image_classification/image_classification.py [hunk 37]
```diff
@@ -1,341 +1,256 @@
...
+        if scheduler_options is not None:
+            config.update(scheduler_options)
+        if use_rec == True:
+            config['use_rec'] = True
+        task = _ImageClassification(config=config)
+        self._classifier = task.fit(train_data, val_data, train_size, random_state)
+        self._fit_summary = task.fit_summary()
+        return self
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply vision/src/autogluon/vision/image_classification/image_classification.py:158]: removed

File: vision/setup.py (hunk 5 above)
Comment (zhreshold) [reply vision/setup.py:51]: v0.9.0, which will be released within a week

File: vision/src/autogluon/vision/image_classification/image_classification.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (Innixma) [reply vision/src/autogluon/vision/image_classification/image_classification.py:116]: What I meant is that you are not lowering the casing when calling `D8D.get()`, I feel like it makes more sense to also pass the lowered casing to the `.get` method. I.E. if `'CIFAR10'`, you are checking if `'cifar10'` in names, and then `train_data = D8D.get('CIFAR10')` when it should probably be `train_data = D8D.get('cifar10')`

File: vision/src/autogluon/vision/image_classification/image_classification.py (hunk 32 above)
[the remark this answers is not in the document]
Comment (Innixma) [reply vision/src/autogluon/vision/image_classification/image_classification.py:133]: `xxx = task.fit()` is confusing now though, because it makes people think that `xxx` and `task` are different things (thats what I thought the first time I read through the website documentation).

File: vision/src/autogluon/vision/image_classification/image_classification.py (hunk 35 above)
[the remark this answers is not in the document]
Comment (Innixma) [reply vision/src/autogluon/vision/image_classification/image_classification.py:142]: I'm not saying to not check for `None`. This is what I meant:

```
        if nthreads_per_trial is not None:
            config['nthreads_per_trial'] = nthreads_per_trial
```

**Comment by szha** [comment]
Job PR-733-56 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-733/56/index.html

**Comment by szha** [comment]
Job PR-733-58 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-733/58/index.html

File: docs/api/autogluon.task.rst (hunk 12 above)
Comment (zhreshold) [reply docs/api/autogluon.task.rst:16]: Fixed

File: docs/tutorials/image_classification/kaggle.md (hunk 11 above)
Comment (zhreshold) [reply docs/tutorials/image_classification/kaggle.md:18]: http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-733/56/tutorials/image_classification/kaggle.html

File: docs/tutorials/object_detection/beginner.md (hunk 10 above)
Comment (zhreshold) [reply docs/tutorials/object_detection/beginner.md]: Fixed

File: docs/tutorials/object_detection/beginner.md (hunk 9 above)
Comment (zhreshold) [reply docs/tutorials/object_detection/beginner.md]: Fixed

**Comment by szha** [comment]
Job PR-733-59 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-733/59/index.html

**Comment by szha** [comment]
Job PR-733-65 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-733/65/index.html

**Comment by zhreshold** [comment]
@Innixma Could you kindly give it another round of review? Doc is here: http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-733/65/index.html

File: vision/src/autogluon/vision/image_prediction/image_prediction.py [hunk 38]
```diff
@@ -0,0 +1,319 @@
...
+            logging.getLogger("ImageClassificationEstimator").setLevel(log_level)
+        task = _ImageClassification(config=config)
+        task._logger.setLevel(log_level)
+        task._logger.propagate = True
+        self._classifier = task.fit(train_data, val_data, 1 - holdout_frac, random_state)
+        self._classifier._logger.setLevel(log_level)
+        self._classifier._logger.propagate = True
+        self._fit_summary = task.fit_summary()
+
+    def predict_proba(self, x):
```
Comment (Innixma) [inline vision/src/autogluon/vision/image_prediction/image_prediction.py:179]: There is still an output format difference here for both predict and predict_proba compared to Tabular. I'm ok with it for this PR but we probably want to align before v0.1 release.

File: vision/src/autogluon/vision/image_prediction/image_prediction.py [hunk 39]
```diff
@@ -0,0 +1,319 @@
...
+        """Dump predictor to disk.
+
+        Parameters
+        ----------
+        file_name : str
+            The file name of saved copy.
+
+        """
+        with open(file_name, 'wb') as fid:
+            pickle.dump(self, fid)
```
Comment (Innixma) [inline vision/src/autogluon/vision/image_prediction/image_prediction.py:279]: Still confused how this doesn't cause issues. I thought mxnet models weren't pickle-able?

**Review by Innixma — APPROVED** [review]
Looks good, still have minor comments regarding docs and broken examples.

Major thing we will want to align on in a future PR is the predict/predict_proba output format, since it currently differs from Tabular.

File: core/setup.py [hunk 40]
```diff
@@ -43,7 +43,7 @@ def create_version_file():
 
 requirements = [
     'numpy>=1.16.0',
-    'scipy>=1.3.3',
+    'scipy>=1.3.3,<1.5.0',
```
[the remark this answers is not in the document]
Comment (zhreshold) [reply core/setup.py:46]: scipy 1.5.4 is causing issue for extra module, non related to this PR, but causing consistent errors in building docs

**Comment by zhreshold** [comment]
Merging this to unblock other PRs, will address docs in a follow up PR.

File: vision/src/autogluon/vision/image_prediction/image_prediction.py (hunk 39 above)
Comment (zhreshold) [reply vision/src/autogluon/vision/image_prediction/image_prediction.py:279]: mxnet models are generally not pickleable due to C pointers to ndarrays(especially cuda arrays, which are not even forkable), I use `__setstate__` and `__getstate__` to bypass the restriction by first use mxnet backend to dump the weights ,load it as binary, and finally remove the network object so the pickled object can be loaded back.
